plugins/inputs/intel_powerstat/intel_powerstat.go

```diff
@@ -229,6 +268,153 @@ func (p *PowerStat) addCPUTemperatureMetric(cpuID string, acc telegraf.Accumulat
 	acc.AddGauge("powerstat_core", fields, tags)
 }
 
+func calculateTurboRatioGroup(coreCounts uint64, msr uint64, group map[int]uint64) {
+	from := coreCounts & 0xFF // value of number of active cores of bucket 1 is written in the first 8 bits. The next buckets values are saved on the following 8-bit sides
+	for i := 0; i < 8; i++ {
+		to := (coreCounts >> (i * 8)) & 0xFF
+		if to == 0 {
+			break
+		}
+		value := (msr >> (i * 8)) & 0xFF
+		// value of freq ratio is stored in 8-bit blocks, and their real value is obtained after multiplication by 100
+		if value != 0 && to != 0 {
+			for ; from <= to; from++ {
+				group[int(from)] = value * 100
+			}
+		}
+		from = to + 1
+	}
+}
+
+func (p *PowerStat) addTurboRatioLimit(socketID string, acc telegraf.Accumulator) {
+	var err error
+	turboRatioLimitGroups := make(map[int]uint64)
+
+	var cpuID = "-1"
+	var model = "-1"
+	for _, v := range p.cpuInfo {
+		if v.physicalID == socketID {
+			cpuID = v.cpuID
+			model = v.model
+		}
+	}
+	if cpuID == "-1" || model == "-1" {
```

Comment:
can we set these to the default string value of "" and test for empty strings at the end rather than assuming -1 is unset?

Reply:
Good point. Code has been corrected